### PR TITLE
feat: Add Azure Monitor input plugin

### LINKED ISSUE #7930: Input Plugin for Azure Monitor
## Feature Request

Currently, there is support for Stackdriver and Amazon Cloudwatch to pull IaaS metrics. Another major IaaS provider in the scene is Microsoft Azure. With the concept of "Multi-Cloud" becoming a norm day by day, this integration of Azure Monitor

### Proposal:
An input plugin that is able to pull metrics from Azure Monitors just like Cloudwatch and Stackdriver plugins that are now present.

### Use case: 
Telegraf is one of the most important and got-to metrics collection agents. And currently when it comes to being able to pull metrics out of Azure Monitors is not out of the box. A plugin for the same would be great.

### PR DESCRIPTION
**Pull request by ShiranAvidov**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #7930

<!-- Finally, include a summary of the code change itself. This
description should tell reviewers how the issues were resolved.

Added an input plugin to gather metrics from Azure Monitor API.

### CODE REVIEW

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 1]
```diff
@@ -0,0 +1,856 @@
...
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type AzureMonitor struct {
+	azureClient *AzureClient
+
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:22]: Usually we put those at the end of the struct, but it's not a mandatory change.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 2]
```diff
@@ -0,0 +1,856 @@
...
+	Resources     []*Resource `toml:"resource"`
+}
+
+type Resource struct {
+	ResourceType string   `toml:"resource_type"`
+	Metrics      []string `toml:"metrics"`
+	Aggregation  []string `toml:"aggregation"`
+}
+
+type AzureClient struct {
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:51]: Do we need to export this structure?

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 3]
```diff
@@ -0,0 +1,856 @@
...
+
+var sampleConfig = `
+# can be found under Overview->Essentials in the Azure portal for your application/service
+subscription_id = "<<SUBSCRIPTION_ID>>"
+# can be obtained by registering an application under Azure Active Directory
+client_id = "<<CLIENT_ID>>"
+# can be obtained by registering an application under Azure Active Directory
+client_secret = "<<CLIENT_SECRET>>"
+# can be found under Azure Active Directory->Properties
+tenant_id = "<<TENANT_ID>>"
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:78]: Those should be prefixed with two spaces to align with other plugins in the produced config.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 4]
```diff
@@ -0,0 +1,856 @@
...
+}
+
+func (am *AzureMonitor) SampleConfig() string {
+	return sampleConfig
+}
+
+// Init is for setup, and validating config.
+func (am *AzureMonitor) Init() error {
+	err := am.checkConfigValidation()
+
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:151]: Please remove this newline as it clutters the code instead of structuring it. The same for other instances below.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 5]
```diff
@@ -0,0 +1,856 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = am.getAccessToken()
+
+	if err != nil {
+		return err
+	}
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:160]: You should add some prefix to the error to easier find the spot if seeing an error message and you can shorten this to
```suggestion
	if err := am.getAccessToken(); err != nil {
		return fmt.Errorf("getting access token failed: %v", err)
	}
```

This has the advantage that `err` is a local variable with the scope of the `if` part...

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 6]
```diff
@@ -0,0 +1,856 @@
...
+
+	return nil
+}
+
+func (am *AzureMonitor) createResourceGroupTargetsFromSubscriptionTargets() error {
+	if len(am.SubscriptionTargets) == 0 {
+		return nil
+	}
+
+	am.Log.Info("Creating resource group targets from subscription targets")
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:247]: Maybe make this a `am.Log.Debug`.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 7]
```diff
@@ -0,0 +1,856 @@
...
+	err = am.getResourceTargetsMetrics()
+
+	if err != nil {
+		return err
+	}
+
+	am.setResourceTargetsAggregation()
+	am.checkTargetsMaximumMetrics()
+
+	am.Log.Info("Total targets: ", len(am.ResourceTargets))
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:183]: Isn't this a debug message?

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 8]
```diff
@@ -0,0 +1,856 @@
...
+	if am.ClientSecret == "" {
+		return fmt.Errorf("client_secret is empty or missing. Please check your configuration")
+	}
+
+	if am.TenantID == "" {
+		return fmt.Errorf("tenant_id is empty or missing. Please check your configuration")
+	}
+
+	if len(am.ResourceTargets) == 0 && len(am.ResourceGroupTargets) == 0 && len(am.SubscriptionTargets) == 0 {
+		return fmt.Errorf("there is no target to collect metrics from. Please check your configuration")
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:206]: Maybe shorten the error message to 
```suggestion
		return fmt.Errorf("no target to collect metrics from in your configuration")
```

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 9]
```diff
@@ -0,0 +1,856 @@
...
+	}
+
+	bodyData, err := unmarshalJSON(body)
+
+	if err != nil {
+		return err
+	}
+
+	for _, value := range bodyData["value"].([]interface{}) {
+		resourceGroup := value.(map[string]interface{})["name"].(string)
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:263]: This might panic if the `bodyData` is not of the expected format (e.g. due to backend changed format or similar). Please check the type-assertions and gracefully return an error in that case to avoid tearing-down telegraf.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 10]
```diff
@@ -0,0 +1,856 @@
...
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
+			}
+		}(target)
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:304]: Is it really necessary to do this in parallel? Remember we are still in `Init()` i.e. in the startup-phase of telegraf... If it's not absolutely necessary to do it in parallel, code can be simplified especially the error handling.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 11]
```diff
@@ -0,0 +1,856 @@
...
+	}
+
+	bodyData, err := unmarshalJSON(body)
+
+	if err != nil {
+		return err
+	}
+
+	for _, value := range bodyData["value"].([]interface{}) {
+		fullResourceID := strings.Split(value.(map[string]interface{})["id"].(string), "/")
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:333]: Same as above for the type-assertion.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 12]
```diff
@@ -0,0 +1,856 @@
...
+	waitGroup.Wait()
+
+	select {
+	case err := <-errChan:
+		return err
+	default:
+	}
+
+	return nil
+}
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:398]: Same here for the parallel execution. If not absolutely necessary, we should do this sequentially to reduce complexity.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 13]
```diff
@@ -0,0 +1,856 @@
...
+			continue
+		}
+
+		metric = NewMetric()
+
+		metric.getMetricName(value.(map[string]interface{}))
+		metric.getMetricFields(data)
+		metric.getMetricTags(bodyData, value.(map[string]interface{}))
+
+		metrics = append(metrics, metric)
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:729]: Can we please directly collect `telegraf.Metric`s here instead of the custom structure definition?

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 14]
```diff
@@ -0,0 +1,856 @@
...
+func (am *AzureMonitor) collectResourceTargetMetrics(body []byte) ([]*Metric, error) {
+	bodyData, err := unmarshalJSON(body)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var metrics []*Metric
+
+	for _, value := range bodyData["value"].([]interface{}) {
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:695]: Please check for the `map` field and the type-assertion to avoid panics here.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 15]
```diff
@@ -0,0 +1,856 @@
...
+			metrics, err := am.collectResourceTargetMetrics(body)
+
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:666]: Please don't do this and instead use `acc.AddError`.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 16]
```diff
@@ -0,0 +1,856 @@
...
+			body, err := am.getTargetResponseBody(apiURL)
+
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:654]: Same as below.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 17]
```diff
@@ -0,0 +1,856 @@
...
+	}
+}
+
+func NewMetric() *Metric {
+	return &Metric{
+		name:   "",
+		fields: make(map[string]interface{}),
+		tags:   make(map[string]string),
+	}
+}
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:506]: I think we don't need to export those functions. And beyond I think you should rather setup the structs at the required place instead of wrapping it in those functions.

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 18]
```diff
@@ -0,0 +1,856 @@
...
+		target.Aggregation = append(target.Aggregation, "Total", "Count", "Average", "Minimum", "Maximum")
+	}
+}
+
+func (am *AzureMonitor) Gather(acc telegraf.Accumulator) error {
+	err := am.refreshAccessToken()
+
+	if err != nil {
+		return err
+	}
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:462]: Shouldn't this be a goroutine started in e.g. `Init()`?

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 19]
```diff
@@ -0,0 +1,856 @@
...
+		return nil, fmt.Errorf("error unmarshalling response body: %v", err)
+	}
+
+	return data, err
+}
+
+func init() {
+	inputs.Add("azure_monitor", func() telegraf.Input {
+		return &AzureMonitor{
+			azureClient: NewAzureClient(),
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:853]: Please do this in `Init()`!

File: plugins/inputs/azure_monitor/azure_monitor_test.go [hunk 20]
```diff
@@ -0,0 +1,774 @@
...
+      }
+	},
+	{
+	  "name": {
+        "value": "Ingress",
+        "localizedValue": "Ingress"
+      }
+	}
+  ]
+}
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor_test.go:220]: Please put this into files (usually located in a `testdata` subdirectory) and load the files in the corresponding tests. This will split the test-data from the test-code and makes the files easier to navigate.

File: plugins/inputs/azure_monitor/azure_monitor_test.go [hunk 21]
```diff
@@ -0,0 +1,774 @@
...
+`
+
+var (
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor_test.go:223]: This looks like you reuse this in the tests. Please don't do this. Setup the tests "standalone" to avoid side-effects.

**Review by srebhan — CHANGES_REQUESTED** [review]
@ShiranAvidov thanks for the interesting PR! Beside the comments in the code, I have a more general comment_:
Please try to simplify the code (e.g. by doing the init-stuff sequentially) and try to maybe reorganize the code to multiple files e.g. for the structures or helper-functions. This would ease the review and later working with the code as the current code base is rather large compared to other plugins.

Looking forward to the next round. :-)

File: plugins/inputs/azure_monitor/azure_monitor.go (hunk 18 above)
Comment (ShiranAvidov) [reply plugins/inputs/azure_monitor/azure_monitor.go:462]: No, each access token is valid for amount of time, and we need to check if we need to create a new one. Without it we won't be able to collect metrics from Azure

Comment (srebhan) [reply plugins/inputs/azure_monitor/azure_monitor.go:462]: Understood. Can you please add this information as a small comment in the code?

**Comment by ShiranAvidov** [comment]
Hi @srebhan,  
I changed the code according your comments.

Happy holidays :)

**Comment by powersj** [comment]
@ShiranAvidov if you could please rebase on master and resolve the conflicts with `go.mod` that would help keep this PR moving.

Thanks!

**Review by srebhan — COMMENTED** [review]
After looking at the code, I wonder why you cannot use the [Azure-SDK for go](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go@v61.1.0+incompatible) instead of directly working with the URLs. If you scroll down you find API packages for [resources/resource groups](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources) as well as [substriptions](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions)...

Any reason you do not use those? They even do have [examples](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/resources/armresources/examples_resources_test.go)...

**Comment by ShiranAvidov** [comment]
Hi @srebhan, I started writing the main feature - collecting metrics, and after finishing I added other ways of getting the resource targets (by resource group and the subscription), so I used the same logic I used for the main feature.
I looked at the sdk, I believe I can use it for getting a list of resource groups and a list of resources inside a resource group.
It looks like the sdk does not have a way to collect metrics using the monitor, so using the api url manually for collecting metrics and getting metric definitions is the best way.

**Comment by srebhan** [comment]
@ShiranAvidov thanks for looking into the SDK. Please also note that they have zillions of packages, so if you miss some functionality maybe also look at the neighboring packages. :-) Thanks again for your effort!

**Comment by ShiranAvidov** [comment]
Hey @srebhan, hope you are doing well.
I finished updating the code + tests.

**Comment by cthiel42** [comment]
Hi @ShiranAvidov. I'm glad to see a PR open for a plugin I was looking for. I played around with this a bit and have a few ideas.

I'm concerned over the timeStamp label causing high cardinality. I'd be curious to know @srebhan opinion on this. For now I'm using fielddrop in my config to get rid of it.

And then a feature idea here (not something that needs to be in this PR, it will probably require a bit of work) could be the ability to backfill data. Right now it looks like every interval it grabs the latest data point. A single API call is capable of pulling a days worth of datapoints. Instead of making an API call every minute to get 1 minute metric resolution, it'd be cool if I were able to set the interval to a greater time and it backfill data for that entire interval. This would save a lot of cost on API calls for when I don't necessarily need the data to be as live as possible. I will admit there's some nuances to a feature like this that probably make it easier said than done. Off the top of my head I can think of interval jitter causing there to be overlapping datapoints.

Thanks for making this plugin. It's simple to use, works well, and I have a use case for something just like this.

File: plugins/inputs/azure_monitor/collector.go [hunk 22]
```diff
@@ -0,0 +1,386 @@
...
+	}
+
+	if metricValue.TimeStamp == nil {
+		return nil
+	}
+
+	metricFields := make(map[string]interface{})
+	metricValueFieldsNum := 1
+
+	metricFields["timeStamp"] = metricValue.TimeStamp.Format("2006-01-02T15:04:05Z07:00")
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/collector.go:211]: Why don't we use this timestamp as the metric timestamp?

File: plugins/inputs/azure_monitor/collector.go [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+
+func getMetricFields(metricValues []*armmonitor.MetricValue) map[string]interface{} {
+	for index := len(metricValues) - 1; index >= 0; index-- {
+		metricFields := getMetricsClientMetricValueFields(metricValues[index])
+		if metricFields == nil {
+			continue
+		}
+
+		return metricFields
+	}
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/collector.go:272]: Is there a reason for this type of iteration? Why do you return the "latest" valid value? A comment would help the reader here.

File: plugins/inputs/azure_monitor/collector.go [hunk 24]
```diff
@@ -0,0 +1,386 @@
...
+
+	resourceGroupName, err := getPartOfMetricID(*metricID, 1, 2, true)
+	if err != nil {
+		return nil, err
+	}
+
+	return resourceGroupName, nil
+}
+
+func getPartOfMetricID(metricID string, partIndex int, partSubPartIndex int, getPartToEnd bool) (*string, error) {
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/collector.go:366]: Why do you work with `string` pointers in all those functions? You could simply return a string...

**Review by srebhan — COMMENTED** [review]
Thanks @ShiranAvidov for your update. I left some comments in the code. Overall I do see an improvement, but I still have the feeling that more than 3.5k lines of code is quite much...

**Comment by ZPascal** [comment]
@ShiranAvidov Cool helpful plugin. I'm looking forward to see the new functionalty in the near future inside the Telegraf.

**Comment by srebhan** [comment]
Honestly speaking an integration would require to dramatically reduce the code size by either using existing libraries or move the inner mechanisms to a "high-level" library hiding the complexity. The present PR has 3.8k of code where around 1.2k is actual functionality and the remaining 2.6k is testing of _internal_ functionality. That's about **three times** the size of most plugins. I really think this should be a library on it's own (or using an existing library if there is any) and a lean telegraf plugin only calls into this _externally_ maintained lib...

**Comment by infa-pavakuma** [comment]
Any Expected time on when this could be merged?

**Comment by srebhan** [comment]
@infa-pavakuma as stated earlier, the PR needs a revamp to reduce the amount of code by either using existing libraries or by moving most of the inner workings to a library. When this happened and the PR is updated I will give it another review. Asking when this will be merged won't speed-up things, but supporting @ShiranAvidov and @ZPascal might... ;-)

**Comment by ShiranAvidov** [comment]
@srebhan @ZPascal thanks for the feedback. @ZPascal can you please elaborate more about your suggestions to reduce the size of the code? Currently I am using Azure library that lets me send requests to the Azure monitor API.

**Comment by srebhan** [comment]
@ShiranAvidov and @ZPascal please keep this going! Just ping me if you need my input...

**Comment by ZPascal** [comment]
@ShiranAvidov I checked your code a bit, and it seems that you have some pre-existing (SDK) methods wrapped in your codebase. I recommend reducing the code size also by separating basic lines of code in a separate/unified library. In the end, it is possible to use this common library within the Telegraf codebase. My offer still stands, if you need assistance with this, we can implement it together.

**Comment by ShiranAvidov** [comment]
Hey @srebhan, 
I reduced the size of the code. The code is in one file - 235 lines!!!

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+		azureClients, err := receiver.CreateAzureClients(am.SubscriptionID, am.ClientID, am.ClientSecret, am.TenantID)
+		if err != nil {
+			return err
+		}
+
+		am.azureClients = azureClients
+	}
+
+	if err := am.setReceiver(); err != nil {
+		return fmt.Errorf("error setting Azure Monitor receiver: %v", err)
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:134]: Can you please use error wrapping here? Same for all other `fmt.Errorf` instances...
```suggestion
		return fmt.Errorf("error setting Azure Monitor receiver: %w", err)
```

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+	for _, target := range am.ResourceTargets {
+		resourceTargets = append(resourceTargets, receiver.NewResourceTarget(target.ResourceID, target.Metrics, target.Aggregations))
+	}
+
+	for _, target := range am.ResourceGroupTargets {
+		resources := make([]*receiver.Resource, 0)
+
+		for _, resource := range target.Resources {
+			resources = append(resources, receiver.NewResource(resource.ResourceType, resource.Metrics, resource.Aggregations))
+		}
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:184]:
```suggestion
		var resources []*receiver.Resource
		for _, resource := range target.Resources {
			resources = append(resources, receiver.NewResource(resource.ResourceType, resource.Metrics, resource.Aggregations))
		}
```

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 27]
```diff
@@ -0,0 +1,235 @@
...
+	}
+
+	targets := receiver.NewTargets(resourceTargets, resourceGroupTargets, subscriptionTargets)
+	rec, err := receiver.NewAzureMonitorMetricsReceiver(am.SubscriptionID, am.ClientID, am.ClientSecret, am.TenantID, targets, am.azureClients)
+	if err != nil {
+		return err
+	}
+
+	am.receiver = rec
+	return nil
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:200]: How about
```suggestion
	var err error
	am.receiver, err = receiver.NewAzureMonitorMetricsReceiver(am.SubscriptionID, am.ClientID, am.ClientSecret, am.TenantID, targets, am.azureClients)
	return err
```

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 28]
```diff
@@ -0,0 +1,235 @@
...
+	for _, target := range am.receiver.Targets.ResourceTargets {
+		am.Log.Debug("Collecting metrics for resource target ", target.ResourceID)
+		waitGroup.Add(1)
+
+		go func(target *receiver.ResourceTarget) {
+			defer waitGroup.Done()
+
+			collectedMetrics, notCollectedMetrics, err := am.receiver.CollectResourceTargetMetrics(target)
+			if err != nil {
+				acc.AddError(fmt.Errorf("%v", err))
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor.go:215]: Simply do
```suggestion
				acc.AddError(err)
```

File: plugins/inputs/azure_monitor/azure_monitor_test.go [hunk 29]
```diff
@@ -0,0 +1,1002 @@
...
+			expectedResourceMetrics = append(expectedResourceMetrics, testMetric1)
+		} else {
+			expectedResourceMetrics = append(expectedResourceMetrics, testMetric2)
+		}
+	}
+
+	err := am.Init()
+	require.NoError(t, err)
+
+	assert.Len(t, am.receiver.Targets.ResourceTargets, 8)
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor_test.go:72]: Please prefer `require` over `assert`.
```suggestion
	require.Len(t, am.receiver.Targets.ResourceTargets, 8)
```

File: plugins/inputs/azure_monitor/azure_monitor_test.go [hunk 30]
```diff
@@ -0,0 +1,1002 @@
...
+	for index := 1; index <= receiver.MaxMetricsPerRequest; index++ {
+		if index <= 10 {
+			expectedResourceMetrics = append(expectedResourceMetrics, testMetric1)
+		} else {
+			expectedResourceMetrics = append(expectedResourceMetrics, testMetric2)
+		}
+	}
+
+	err := am.Init()
+	require.NoError(t, err)
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor_test.go:70]:
```suggestion
	require.NoError(t, am.Init())
```

**Review by srebhan — COMMENTED** [review]
Wow that is a __HUGE__ improvement @ShiranAvidov! Much appreciated.

I have some more comments in the code, nothing too big though. Now let's talk about the tests. Do you think it would be possible to move the test-data to e.g. JSON files and load/unmarshal them during the test? This would make the test-logic more obvious and reduces the code-size further...

**Comment by srebhan** [comment]
Hey @ShiranAvidov, is there a chance you can address my comments?

**Comment by ShiranAvidov** [comment]
Hey @srebhan, sorry had a lot of other work to do. I think I can move the data into json files, but I will still have to use the List methods (the load/unmarshal will be in the List methods).

File: plugins/inputs/azure_monitor/azure_monitor.go [hunk 31]
```diff
@@ -0,0 +1,235 @@
...
+// Init is for setup, and validating config.
+func (am *AzureMonitor) Init() error {
+	if am.azureClients == nil {
+		azureClients, err := receiver.CreateAzureClients(am.SubscriptionID, am.ClientID, am.ClientSecret, am.TenantID)
+		if err != nil {
+			return err
+		}
+
+		am.azureClients = azureClients
+	}
```
[the remark this answers is not in the document]
Comment (ShiranAvidov) [reply plugins/inputs/azure_monitor/azure_monitor.go:131]: The if condition is for the tests. In the tests functions azureClients is being set with mock object, and in regular run it is nil.

**Comment by ShiranAvidov** [comment]
Hey @srebhan, finished fixing the code. I moved all test data to json/toml files as you requested :)

File: plugins/inputs/azure_monitor/azure_monitor.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/inputs/azure_monitor/azure_monitor.go:130]: Please avoid this. Set the receiver in those tests maybe after the `Init()` or provide a `createClient()` function that will be overwritten in the tests.

**Review by srebhan — COMMENTED** [review]
@ShiranAvidov thanks for the nice update! Please fix the one issue in `Init()` as the test should not dictate the flow of the code. If necessary, please abstract setting the "client" using a function, modify the client in the after calling `Init()` or split the `Init()` into setting up the client and the rest and call what you need in the test.

**Comment by ShiranAvidov** [comment]
@srebhan finished fixing the code.

**Comment by ZPascal** [comment]
Hi @srebhan, is there the possibility to merge the PR?

File: plugins/inputs/all/all.go [hunk 32]
```diff
@@ -1 +1,225 @@
...
+	_ "github.com/influxdata/telegraf/plugins/inputs/win_perf_counters"
+	_ "github.com/influxdata/telegraf/plugins/inputs/win_services"
+	_ "github.com/influxdata/telegraf/plugins/inputs/wireguard"
+	_ "github.com/influxdata/telegraf/plugins/inputs/wireless"
+	_ "github.com/influxdata/telegraf/plugins/inputs/x509_cert"
+	_ "github.com/influxdata/telegraf/plugins/inputs/xtremio"
+	_ "github.com/influxdata/telegraf/plugins/inputs/zfs"
+	_ "github.com/influxdata/telegraf/plugins/inputs/zipkin"
+	_ "github.com/influxdata/telegraf/plugins/inputs/zookeeper"
+)
```
Comment (srebhan) [inline plugins/inputs/all/all.go:225]: We changed the way we import plugins to be able to selectively build plugins. So can you please rename this file to `azure_monitor.go` and change the content to
```suggestion
//go:build !custom || inputs || inputs.azure_monitor
package all

import _ "github.com/influxdata/telegraf/plugins/inputs/azure_monitor" // register plugin
```

File: plugins/inputs/azure_monitor/azure_monitor_test.go [hunk 33]
```diff
@@ -0,0 +1,954 @@
...
+	}
+
+	return nil, fmt.Errorf("resouce group was not found")
+}
+
+func (mamdc *mockAzureMetricDefinitionsClient) List(
+	_ context.Context,
+	resourceID string,
+	_ *armmonitor.MetricDefinitionsClientListOptions) (armmonitor.MetricDefinitionsClientListResponse, error) {
+
```
Comment (srebhan) [inline plugins/inputs/azure_monitor/azure_monitor_test.go:114]: Remove the empty line.

**Review by srebhan — COMMENTED** [review]
Thanks @ShiranAvidov for the update. Just one empty line and changing the `all.go` file. Please make sure to rename it to not conflict. Last but not least, please fix the conflict in `go.mod` and we are good to go.

**Comment by ShiranAvidov** [comment]
@srebhan done.

**Comment by ZPascal** [comment]
@srebhan Are there still open points from your side? Is the PR mergeable?

**Comment by ZPascal** [comment]
@srebhan Any news from you side?

**Comment by ZPascal** [comment]
@powersj Is there the possibility to rebase and merge it? We want to consume the new functionality.

**Review by srebhan — APPROVED** [review]
Looks good to me. Sorry for the late reply... Thanks for all your effort @ShiranAvidov and @ZPascal!

**Comment by srebhan** [comment]
@ZPascal can you please resolve the remaining conflict?

**Comment by powersj** [comment]
I just pushed to resolve the go.sum conflict, let's see if tests pass

**Comment by powersj** [comment]
> please run go mod tidy and check in changes, you might have to use the same version of Go as the CI


Doesn't look like it, so still need you to push an update after running `go mod tidy`

**Comment by powersj** [comment]
@srebhan can you re-reivew, I've cleaned up linter issues, ran make tidy, and moved the sample config out

**Review by powersj — APPROVED** [review]

**Comment by ZPascal** [comment]
@powersj Thank you for handling the merge conflicts. Also, from side thank you, @ShiranAvidov @srebhan @powersj for the review, the feature and the invested work to deliver the functionality.
